lib/WebService/Async/Onfido/RateLimiter.pm

```diff
@@ -0,0 +1,158 @@
+use mro;
+
+use parent qw(IO::Async::Notifier);
+
+=head1 NAME
+
+WebService::Async::CustomerIO::RateLimitter - This class provide possibility to limit amount
+of request in time interval
+
+=head1 SYNOPSIS
+
+   use IO::Async::Loop;
+   use WebService::Async::Onfido::RateLimiter;
+   my $loop = IO::Async::Loop->new;
+   $loop->add(
+       my $limiter = WebService::Async::Onfido::RateLimiter->new(
+           interval => 5,
+           limit    => 3
+       ));
+   my $f = $limiter->acquire->then(sub{print "doing "});
+   $loop->run;
+
+=head1 DESCRIPTION
+
+=cut
+
+sub _init {
+    my ($self, $args) = @_;
+    for my $k (qw(limit interval)) {
+        die "Missing required argument: $k" unless exists $args->{$k};
+        die "Invalid value for $k: $args->{$k}" unless int($args->{$k}) > 0;
+        $self->{$k} = delete $args->{$k};
+    }
+
+    $self->{queue}   = [];
+    $self->{counter} = 0;    ## TODO remove ?
+
+    return $self->next::method($args);
+}
+
+=head2 interval
+
+=cut
+
+sub interval { return shift->{interval} }
+
+=head2 limit
+
+=cut
+
+sub limit { return shift->{limit} }
+
+=head2 is_limited
+
+=cut
+
+sub is_limited {
+    my $self = shift;
+    # the number of slots is less.
+    return scalar($self->{queue}->@*) >= $self->limit
+        # the item of [-$limit] is not ready
+        && (
+        !($self->{queue}[-$self->limit]->is_ready)
+        #  or is ready but passed no more than interval seconds
+        || (time() - $self->{queue}[-$self->limit]->get < $self->interval)) ? 1 : 0;
+}
+
+=head2 acquire
+
+Method checks availability for free slot.
+It returns future, when slot will be available, then future will be resolved.
+
+=cut
+
+sub acquire {
+    my $self = shift;
+    my $queue = $self->{queue};
+
+    # if the queue is not filled enough, then it is available
+    if (scalar $queue->@* < $self->limit) {
+        push @$queue, Future->done(time());
+        return $queue->[-1];
+    }
+
+    # else , the current request's available time is the execution timestamp of the last $limit request push the interval
+    my $item     = $queue->[-$self->limit];
+    my $interval = $self->interval;
+    my $loop     = $self->loop;
+    push @$queue, $item->then(
+        sub {
+            my $item_time = shift;
+            # execute after
+            my $after = $item_time + $interval - time();
+            $loop->delay_future(after => $after)->then(
+                sub {
+                    # remove old slots before current slot
+                    for (0 .. $#$queue) {
+                        last unless $queue->[0]->is_ready;
+                        # if the slot too old
+                        if (time() - $queue->[0]->get > $interval) {
+                            shift @$queue;
+                        } else {
+                            last;
+                        }
+                    }
+                    Future->done(time());
+                });
+        });
+    return $queue->[-1];
+}
```

Comment:
If I'm not mistaken, it works incorrectly:  whenever we have `limit` items in  queue, we stop and wait `interval`.

Reply:
also there is no `fifo` guarantee here

Reply:
`push` is done here: https://github.com/binary-com/perl-WebService-Async-Onfido/pull/17/files#diff-08a10dd60098b634de31658c820a222bR114

Reply:
> If I'm not mistaken, it works incorrectly: whenever we have `limit` items in queue, we stop and wait `interval`.

Whenver we have `limit` items in the queue, we will execute after that `$queue[-$limit]` items executed and wait for interval seconds. 
For example, `$queue[-$limit]` is executed at timestamp 5, and current timestamp is 10, and the interval is 10, then it will be executed at timestamp 15,

Reply:
yes, you're right it should work. Initially i miss understand  the idea.